File: 491.递增子序列.cpp
/*
 * @lc app=leetcode.cn id=491 lang=cpp
 *
 * [491] 递增子序列
 */

#include <vector>
#include <unordered_set>

using namespace std;

// @lc code=start
class Solution {
public:
    vector<vector<int>> ans;
    vector<int> temp;
    unordered_set<int> s;
    int n;

    vector<vector<int>> findSubsequences(vector<int> &nums)
    {
        n = nums.size();
        for (int i = 0; i < (1 << n); i++)
        {
            findSubsequences(i, nums);
            int hashValue = getHash(263, int(1E9) + 7);
            if (check() && !s.count(hashValue))
            {
                ans.push_back(temp);
                s.insert(hashValue);
            }
        }
        return ans;
    }

    bool check()
    {
        for (size_t i = 1; i < temp.size(); i++)
        {
            if (temp[i] < temp[i - 1])
            {
                return false;
            }
        }

        return temp.size() >= 2;
    }

    int getHash(int base, int mod)
    {
        int hashValue = 0;
        for (const auto &x : temp)
        {
            hashValue = (1LL * hashValue * base + x + 101) % mod;
        }
        return hashValue;
    }

    void findSubsequences(int mask, vector<int> &nums)
    {
        temp.clear();
        for (size_t i = 0; i < n; i++)
        {
            if (mask & 1)
            {
                temp.push_back(nums[i]);
            }
            mask >>= 1;
        }
    }
};
// @lc code=end

File: 491.递增子序列.java
/*
 * @lc app=leetcode.cn id=491 lang=java
 *
 * [491] 递增子序列
 */

import java.util.ArrayList;
import java.util.List;

// @lc code=start
class Solution {
    List<List<Integer>> ans = new ArrayList<List<Integer>>();
    List<Integer> temp = new ArrayList<Integer>();

    public List<List<Integer>> findSubsequences(int[] nums) {
        dfs(0, Integer.MIN_VALUE, nums);
        return ans;
    }

    private void dfs(int curIndex, int lastNum, int[] nums) {
        if (curIndex == nums.length) {
            if (temp.size() >= 2) {
                ans.add(new ArrayList<Integer>(temp));
            }
            return;
        }

        // 1. If current number is bigger, choose or skip it
        // 2. If current number is equal, choose it
        // 3. If current number is smaller, skip it
        if (nums[curIndex] >= lastNum) {
            temp.add(nums[curIndex]);
            dfs(curIndex + 1, nums[curIndex], nums);
            temp.remove(temp.size() - 1);
        }
        if (nums[curIndex] != lastNum) {
            dfs(curIndex + 1, lastNum, nums);
        }
    }
}
// @lc code=end
